libs/logging/src/base_logger.ts

```diff
@@ -25,18 +27,118 @@ export const CLIENT_SIDE_LOG_SOURCES = [
   LogSource.VxPollBookFrontend,
 ];
 
+/**
+ * The maximum number of repeated log lines before we emit a log line with repeat details, even if
+ * a repeat sequence hasn't ended yet.
+ */
+const MAX_REPEAT_COUNT_BEFORE_REPEAT_LOG_EMIT = 100;
+
+/**
+ * The maximum number of seconds between the first and last repeat of a log line before we emit a
+ * log line with repeat details, even if a repeat sequence hasn't ended yet.
+ */
+const MAX_SECONDS_BETWEEN_FIRST_AND_LAST_REPEAT_BEFORE_REPEAT_LOG_EMIT = 60;
+
 const debug = makeDebug('logger');
 
 export interface LogData extends Dictionary<string | boolean | number> {
   message?: string;
   disposition?: LogDisposition;
 }
 
+/**
+ * A utility class for tracking and coalescing repeated log lines to reduce log spam while still
+ * maintaining fidelity.
+ */
+class RepeatLogTracker {
+  private trackedLogLine?: LogLine;
+  private repeatCount: number = 0;
+  private firstRepeatAt?: Date;
+  private lastRepeatAt?: Date;
+
+  /**
+   * Checks whether the incoming log line is a repeat of the currently tracked log line and returns
+   * stats on the currently tracked log line.
+   */
+  track(incomingLogLine: LogLine): {
+    /**
+     * Whether the incoming log line is a repeat of the currently tracked log line.
+     */
+    isRepeat: boolean;
+    /**
+     * The repeat count of the currently tracked log line, including the incoming log line.
+     */
+    repeatCount: number;
+    /**
+     * The time between the first and last repeat of the currently tracked log line.
+     */
+    secondsBetweenFirstAndLastRepeat?: number;
+  } {
+    const isRepeat =
+      // Perform cheap checks before kicking off the more expensive deep equality check
+      incomingLogLine.eventId === this.trackedLogLine?.eventId &&
+      incomingLogLine.message === this.trackedLogLine?.message &&
+      deepEqual(incomingLogLine, this.trackedLogLine);
+    if (isRepeat) {
+      this.recordRepeat();
+    }
+    return {
+      isRepeat,
+      repeatCount: this.repeatCount,
+      secondsBetweenFirstAndLastRepeat: !this.firstRepeatAt
+        ? undefined
+        : (assertDefined(this.lastRepeatAt).getTime() -
+            this.firstRepeatAt.getTime()) /
+          1000,
+    };
+  }
+
+  /**
+   * Emits a log line with repeat details about the currently tracked log line if applicable and
+   * restarts tracking using the provided log line as the new log line to track.
+   */
+  flushAndRestartTracking(
+    log: (logLine: LogLine) => void,
+    logLineToTrack: LogLine
+  ): void {
+    if (this.repeatCount > 0) {
+      log(this.prepareLogLineWithRepeatDetails());
+    }
+    this.trackedLogLine = logLineToTrack;
+    this.repeatCount = 0;
+    this.firstRepeatAt = undefined;
+    this.lastRepeatAt = undefined;
+  }
+
+  private recordRepeat(): number {
+    this.repeatCount += 1;
+    if (this.repeatCount === 1) {
+      this.firstRepeatAt = new Date();
+    }
+    this.lastRepeatAt = new Date();
```

Comment:
Should these timestamps be the same for the first repetition of a log? Maybe doesn't matter in real world usage

Reply:
Oh yeah good catch! They happened to be exactly the same in my testing but lemme make sure they really are the same